### ISSUE TITLE
2.5.22294: UnitImageFactory#lambda$getImageDimensions$2:326 - games.strategy.triplea.image.MissingImageException

### ISSUE THREAD

**Issue by tripleabuilderbot** [triplea-game/triplea#8102]
## Map
world_war_ii_classic

## Log Message
Missing image: UnitImageFactory.ImageKey(player=PlayerId named:British, type=UnitType{name=aaGun}, damaged=false, disabled=false)

## TripleA Version
2.5.22294

## Java Version
11.0.6

## Operating System
Windows 10

## Memory
Heap utilization statistics [MB]
Used Memory: 220
Free memory: 1758
Total memory: 1979
Max memory: 1979

## Stack Trace
```
Exception: games.strategy.triplea.image.MissingImageExceptionMissing image: UnitImageFactory.ImageKey(player=PlayerId named:British, type=UnitType{name=aaGun}, damaged=false, disabled=false)
java.lang.Exception
	at games.strategy.triplea.image.UnitImageFactory.lambda$getImageDimensions$2(UnitImageFactory.java:326)
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at games.strategy.triplea.image.UnitImageFactory.getImageDimensions(UnitImageFactory.java:326)
	at games.strategy.triplea.ui.unit.scroller.AvatarPanelFactory.createUnitStackImage(AvatarPanelFactory.java:112)
	at games.strategy.triplea.ui.unit.scroller.AvatarPanelFactory.buildPanel(AvatarPanelFactory.java:86)
	at games.strategy.triplea.ui.unit.scroller.UnitScroller.lambda$drawUnitAvatarPane$6(UnitScroller.java:225)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)


```

**Comment by bacrossland** [comment]
I just ran into this issue. I opened the attached World War II Global 1940 Balanced Mod 3 with Combat Move First game
[Turn4_slow9.tsvg.zip](https://github.com/triplea-game/triplea/files/7207818/Turn4_slow9.tsvg.zip)
 and applied the downloaded battlemap_skin. 

Since I am on the German turn, as soon as I move the mouse over a territory or sea zone with German units in it, the error appears. In the captured error on the ticket, the issue is happening to a British player.

**Comment by bacrossland** [comment]
If you change the skin back to original, the error continues to happen. You have to shut the game down and restart it to clear the error from happening.

**Comment by bacrossland** [comment]
Error only happens in a phase in which the UnitScroller (lower right of the UI screen) is active. Images for the units are missing from there whereas in a game unaffected by this error the units are present.

**Comment by Cernelius** [comment]
(Why did the "Stale" label not go away?)

**Comment by DanVanAtta** [comment]
@Cernelius  stalebot is disabled

**Comment by RoiEXLab** [comment]
@bacrossland I'm not familiar with the details of all maps, but usually there kind of errors occur because some files are actually missing. However both repos you listed contain the `unit/Germans/infantry.png` file which is unexpected.

`ResourceLoader` basically delegates most of its logic to `URLClassLoader` to be able to debug this you might want to try using isolate code snippets, something like this:
```java
new URLClassLoader(new URL[]{ new URL("file:/path/to/map/folder")}).resources("units/Germans/infantry.png").orElse(null);
```

If you then analyse what kind of parameters are returning `null` you might be able to find out what exactly is happening, but I don't know enough about this particular code to make a educated guess here

**Comment by DanVanAtta** [comment]
2.6 handling of skins is incomplete and different from 2.5

**Comment by Cernelius** [comment]
@DanVanAtta Where is the documentation (fully) explaining how non-original skins are supposed to be handled via TripleA 2.6? As far as myself is concerned, I'm really only interested in knowing all the differences from 2.5.

I remember that the change was that non-original skins would be within the map itself (instead of being map-skins). Is this the only difference or are there other changes? Myself, I would strongly advocate for also allowing every item to be optional, the program defaulting to using the original skin wherever a non-original one is missing the item (so that there may be no need of duplicating assets).

Also since every original skin, which any map has, is a skin too, am I correct to assume that by "handling of skins" you actually mean "handling of non-original skins" (so far known as map-skins), or do you mean that also original skins (that is, every map in TripleA) is undergoing the implied changes (whatever they are)?

On the matter, I believe the "battlemap_skin" is the most important map-skin of TripleA, so it is likely necessary for it to work fine if map-skins are supposed still to work at all (as I understand they do).

**Comment by bacrossland** [comment]
> 2.6 handling of skins is incomplete and different from 2.5

@DanVanAtta is 2.5 open for taking patches?

@Cernelius the fix I used to load it was my own doing based on just seeing how the code worked. If the 2.6 skin implementation is incomplete the final result might not need a file like that being added for older map skins. What I can tell you is that once it loaded, the only issue I noticed is the error this thread discusses. Everything else appeared to be working as normal.
